### ISSUE TITLE
filesystem provider should trigger "success" event

### ISSUE THREAD

**Issue by technicallyfeasible** [strongloop/loopback-component-storage#58]
The S3 storage component of pkgcloud wraps the upload stream into a proxy and forwards the "uploaded" event as "success" as well as the "data" and "error" events.
The filesystem provider that comes with loopback-component-storage returns the fs.createWritableStream which triggers the usual filesystem events. That means that it only triggers "finish" which triggers too early for S3 storage.

This means that at the moment it is not possible to write a truly provider independent upload code.

**Comment by sayden** [comment]
I have also faced the same problem. Using the filesystem, you can open the lib/storage-handler.js and within the upload method add an event (aroung L123) when the writing on disk has finished. I managed to write a workaround like this

```javascript
// on node_modules/loopback-component-storage/lib/storage-handler   .upload method
var storageHandler = this;

writer.on('finish', function(){
  storageHandler.notifyWriteFinished(file.name, container);
});

//Within the same file
exports.notifyWriteFinished = function(fileName, container){
  var filePath = '/tmp/storage/' + container + '/' + fileName

  var app = require('../../../server/server');
  var models = app.models();
  var Container = _.find(models, function(model){
    // In my case, the model was named 'Container'
    if(model.modelName === 'Container'){
      return model;
    }
  });

  // Inside the model, I wrote an uploadCompleted method
  Container.uploadCompleted(filePath, container);
}

//on common/models/[model].js
  var ctx, next, upload;
  /**
   * Hook executed when a file has been uploaded
   */
  Container.afterRemote('upload', function (context, _upload, _next) {
     
    var _context = loopback.getCurrentContext();
    ctx = _context;
    next = _next;
    upload = _upload;
  });

  
  Container.uploadCompleted = function (filePath, containerName) {
    //Do whatever you need
   
   //Add something to the json result
   upload.result.fields.aParam = 'myParam';

  //Finally call the middleware
  next();
  });
```

A fix is need in here because the afterRemote method is called too early.

**Comment by PeterClough** [comment]
Yeah same problem here.

Hey @sayden  I've tried to implement what you've suggested and have done it exactly how you said but I get this error.

```/Users/peterclough/Intranet/myApp/node_modules/loopback-component-storage/lib/storage-handler.js:126
      storageHandler.notifyWriteFinished(file.name, container);
                     ^
TypeError: Object #<IncomingForm> has no method 'notifyWriteFinished'
    at WriteStream.<anonymous> (/Users/peterclough/Intranet/myApp/node_modules/loopback-component-storage/lib/storage-handler.js:126:22)
    at WriteStream.emit (events.js:117:20)
    at finishMaybe (_stream_writable.js:359:12)
    at afterWrite (_stream_writable.js:279:5)
    at onwrite (_stream_writable.js:269:7)
    at WritableState.onwrite (_stream_writable.js:96:5)
    at fs.js:1688:5
    at /Users/peterclough/Intranet/myApp/node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31
```

I did put this in the same file as you suggested.

```
//Within the same file
exports.notifyWriteFinished = function(fileName, container){
  var filePath = '/tmp/storage/' + container + '/' + fileName

  var app = require('../../../server/server');
  var models = app.models();
  var Container = _.find(models, function(model){
    // In my case, the model was named 'Container'
    if(model.modelName === 'Container'){
      return model;
    }
  });
```

Any ideas on what I've done wrong?

**Comment by sayden** [comment]
I forgot to mention explicitely (but it's written in the example) that an event notifier must be added in the upload method

```javascript
// on node_modules/loopback-component-storage/lib/storage-handler   .upload method
var storageHandler = this;

writer.on('finish', function(){
  storageHandler.notifyWriteFinished(file.name, container);
});
```

Maybe that's the problem. I have correct the workaround

**Comment by PeterClough** [comment]
Yeah I put it in that's the line thats causing the error and its bang on line 123.

```
    var uploadParams = {container: container, remote: file.name, contentType: file.type};
    if (file.acl) {
      uploadParams.acl = file.acl;
    }
    var writer = provider.upload(uploadParams);

    writer.on('error', function(err) {
      self._error(err);
    });

    // on node_modules/loopback-component-storage/lib/storage-handler   .upload method
    var storageHandler = this;

    writer.on('finish', function(){
      storageHandler.notifyWriteFinished(file.name, container);
    });

    var endFunc = function () {
      self._flushing--;
      var values = files[part.name];
      if (values === undefined) {
        values = [file];
        files[part.name] = values;
      } else {
        values.push(file);
      }
      self.emit('file', part.name, file);
      self._maybeEnd();
    };````

**Comment by richardpringle** [comment]
@technicallyfeasible, @sayden, @PeterClough

I see that this issue is just about a year old but I am wondering if you guys were able to check and see whether or not the above two commits fixed the issue?

**Comment by sayden** [comment]
@richardpringle in our case, we continued with the solution I wrote above but we didn't update the server to avoid possible problems.

It was a POC and finally we started a new version of the server without Strongloop, sorry!

**Comment by PeterClough** [comment]
@richardpringle Yes Raymond fixed it in 1.5.0 and it works fine for me.
